pkg/controller/eventinstance/truncate.go

```diff
@@ -0,0 +1,164 @@
+package eventinstance
+
+import (
+	"context"
+	"errors"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/acorn-io/baaah/pkg/router"
+	internalv1 "github.com/acorn-io/runtime/pkg/apis/internal.acorn.io/v1"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func Truncate(ctx context.Context, c kclient.Client, capacity int) router.HandlerFunc {
+	t := &truncator{
+		client:   c,
+		capacity: capacity,
+		tracking: make(map[kclient.ObjectKey]struct{}),
+		mu:       new(sync.Mutex),
+	}
+	var once sync.Once
+
+	return func(req router.Request, _ router.Response) error {
+		once.Do(func() {
+			// Lazy start the truncator.
+			// This ensures it's started after the router cache has synced
+			go func() {
+				if err := t.start(ctx); err != nil {
+					logrus.WithError(err).Warn("Exiting eventinstance truncator")
+				}
+			}()
+		})
+
+		eventKey := router.Key(req.Namespace, req.Name)
+		if req.Object != nil {
+			// Add or Update, add to the running count of events
+			t.track(eventKey)
+		} else {
+			// Delete, remove from the running count of events
+			t.forget(eventKey)
+		}
+
+		return nil
+	}
+}
+
+type truncator struct {
+	client   kclient.Client
+	capacity int
+	tracking map[kclient.ObjectKey]struct{}
+	period   time.Duration
+	trigger  chan struct{}
+	mu       *sync.Mutex
+}
+
+func (t *truncator) start(ctx context.Context) error {
+	timer := time.NewTicker(t.period)
```

Comment:
I don't think we need to do this periodically. We probably just want a "cooldown" between runs of `r.truncate`